### PR TITLE
Phase-2 NN Calo Tau

### PR DESCRIPTION
**Pull request by jonamotta**
#### PR description:

This PR uploads the Phase-2 NN Calo Tau producer and emulator - The TauMinator.

The producer/eulator reconstructs and identifies tau candidates at L1 Trigger with a CNN.
The output is a standard BXVector filled with tau objects.
For simplicity, the dataformat of taus is kept the same as the one of Phase-1

The producer/emulator depends on `.pb` files containing the graphs of the various NNs, which can be found in this PR [#17](https://github.com/cms-l1t-offline/L1Trigger-L1TCalorimeter/pull/17).

Some details about the algorithm and the performance evaluated privately can be found in [this presentation](https://indico.cern.ch/event/1290402/#8-phase-2-nn-calo-taus) and [this presentation](https://indico.cern.ch/event/1290342/#16-status-and-plans-of-the-nn)

**INFO** : IN THE EMULATOR THE NNs ARE STILL APPLIED USING THE TENSORFLOW SOFTWARE
The implementation of full emulation via hls4ml is ongoing (it has already been shown in other contexts that tensorflow  softwrae and full emulation are very close to each other)

#### PR validation:

The code has been validated with the use of the following two standard commands:
`scram build code-checks`
`scram build code-format`
and in both cases no error or warning was prompted by the code in this PR.

#### WP definition and usage:

The WPs on the tau identification are defined based on the total efficiency of the algorithm, i.e. 90/95/99%.
The WPs cut is applied directly in the code and a "quality flag" is saved in the produced taus.
For simplicity, the dataformat of taus is kept the same as the one of Phase-1.
The "quality flag" in the output tau is stored in the hardware quality property of the tau, and can be accessed as:
```
const l1t::Tau& tau = ...;
tau.hwQual();
```
The WP-hwQual correspondance is:
- 90% --> 3
- 95% --> 2
- 99% --> 1
- anything else --> 0

The results in the presentations linked above are all obtained using the 90% WP, which corresponds to asking `tau.hwQual()==3`; to use the other two WP the requirement would be `tau.hwQual()>=2` and `tau.hwQual()>=1` for 95 and 99% WP, respectively.

#### Associated PR to CMSSW master
[PR #42840](https://github.com/cms-sw/cmssw/pull/42840)

### CODE REVIEW

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 1 files that did not meet formatting requirements:
 - L1Trigger/L1CaloTrigger/plugins/L1NNCaloTauProducer.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by jonamotta** [comment]
I have re-run the `scram b code-format` and now it actually did re-format the code.
Not sure why this did not happen the first time I ran it. Sorry for that.

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no files with code format issues!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

File: L1Trigger/L1CaloTrigger/python/L1NNCaloTauProducer_cfi.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import FWCore.ParameterSet.Config as cms
+
+L1NNCaloTauProducer = cms.EDProducer("L1NNCaloTauProducer",
```
Comment (EmyrClement) [inline L1Trigger/L1CaloTrigger/python/L1NNCaloTauProducer_cfi.py]: The naming convention typically used now is to start with `l1t`, so something like `l1tNNCaloTau` would be a better label.

**Review by EmyrClement — COMMENTED** [review]
A few comments from checking out your code, and validating with the menu ntuples.

Can you add your producer to the [default L1 sequence](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-1252patch1/L1Trigger/Configuration/python/SimL1Emulator_cff.py) and the output products to the [event content](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-1252patch1/L1Trigger/Configuration/python/L1Trigger_EventContent_cff.py)?

Can you also provide a WP (or two) for the NN ID that I should try, and a recipe on how to apply it based on the variables in the output product?

**Comment by jonamotta** [comment]
Addressed the comments received both here and offline.
(Putting WP explanation in the description of the PR)

File: L1Trigger/L1CaloTrigger/python/L1NNCaloTauProducer_cfi.py (hunk 1 above)
Comment (jonamotta) [reply L1Trigger/L1CaloTrigger/python/L1NNCaloTauProducer_cfi.py]: Done

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no files with code format issues!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 1 files that did not meet formatting requirements:
 - L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by aloeliger** [comment]
@jonamotta We are starting to prepare for the next round of phase 2 development, and the IB branch has been swapped to phase 2. Could we get you to rebase this branch onto CMSSW_13_3_0_pre3 to sort out the merge issues?

**Comment by mcepeda** [comment]
Hi - just a ping on this @jonamotta. Is this PR still active? Has it been superceded?

**Comment by jonamotta** [comment]
Closing this PR at the request of @EmyrClement because of the mess of commits that happened.
Therefore, this PR is superseded by [PR#1197](https://github.com/cms-l1t-offline/cmssw/pull/1197)

**Pull request closed without merging**